benchmarks/bench_gap_divergence.py

```diff
@@ -0,0 +1,201 @@
+  - Set RNG seed
+  - Set `min_iter` to 5 (`= max_iter`)
+  - Set `max_iter_e_step` to values from 1 to 20
+- Before the end of each loop, save the following:
+  - `W_change`
+  - `score`, from ``self.score(X)``
+  - `W_`
+  - `A_`
+  - `B_`
+- For the visualization, extract from the matrices their:
+  - mean
+  - determinant (`scipy.linalg.det`)
+  - singular values (`s` of `scipy.linalg.svd`)
+
+Date: June 12th 2023
+Commit: 7fc998dcafe12764c1c8ddf5b9f271868c720801
+"""
+
+import scipy as sp
+import numpy as np
+import pandas as pd
+
+from typing import List, Dict, Union
+from skrub.datasets import fetch_employee_salaries
+from argparse import ArgumentParser
+from skrub._gap_encoder import (
+    GapEncoder,
+    GapEncoderColumn,
+    _multiplicative_update_h,
+    _multiplicative_update_w,
+    batch_lookup,
+)
+
+from utils import monitor, default_parser, find_result
+
+
+class ModifiedGapEncoderColumn(GapEncoderColumn):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.benchmark_results_: List[Dict[str, Union[np.ndarray, float]]] = []
+
+    def fit(self, X, y=None):
+        # Copy parameter rho
+        self.rho_ = self.rho
+        # Check if first item has str or np.str_ type
+        assert isinstance(X[0], str), "Input data is not string. "
+        # Make n-grams counts matrix unq_V
+        unq_X, unq_V, lookup = self._init_vars(X)
+        n_batch = (len(X) - 1) // self.batch_size + 1
+        score = self.score(X)
+        del X
+        # Get activations unq_H
+        unq_H = self._get_H(unq_X)
+
+        for n_iter_ in range(self.max_iter):
+            # Loop over batches
+            for i, (unq_idx, idx) in enumerate(batch_lookup(lookup, n=self.batch_size)):
+                if i == n_batch - 1:
+                    W_last = self.W_.copy()
+                # Update activations unq_H
+                unq_H[unq_idx] = _multiplicative_update_h(
+                    unq_V[unq_idx],
+                    self.W_,
+                    unq_H[unq_idx],
+                    epsilon=1e-3,
+                    max_iter=self.max_iter_e_step,
+                    rescale_W=self.rescale_W,
+                    gamma_shape_prior=self.gamma_shape_prior,
+                    gamma_scale_prior=self.gamma_scale_prior,
+                )
+                # Update the topics self.W_
+                _multiplicative_update_w(
+                    unq_V[idx],
+                    self.W_,
+                    self.A_,
+                    self.B_,
+                    unq_H[idx],
+                    self.rescale_W,
+                    self.rho_,
+                )
+
+                if i == n_batch - 1:
+                    # Compute the norm of the update of W in the last batch
+                    W_change = np.linalg.norm(self.W_ - W_last) / np.linalg.norm(W_last)
+
+            print(self.W_.shape, self.A_.shape, self.B_.shape)
+
+            self.benchmark_results_.append(
```

Comment:
I've improved `monitor` to automatically save intermediary results, thanks for the suggestion :)
For the first point, I've left it before the break so that we can see with greater detail what's going on (and to be honest, the matrices aren't that large for `employee_salaries`, only ~3700 x 10 each), we might reconsider this when working with the other datasets!

Reply:
For the last part, it's just that we're going to have mostly duplicated points, so it will be more work for you to clean your data for plotting ;)